---
title: "Assignment 4"
author: "Krishna Kalyan"
output: html_document
---

```{r,message=F,warning=F,echo=F}
library(HardyWeinberg)
library(LDheatmap)
library(genetics)
library(ggplot2)
```

1. The file ABO-CHB.rda contains genotype information of individuals of a Chinese population of unrelated individuals. The genotype information concerns SNPs the ABO bloodgroup region, located on chromosome number 9. The file contains genotype information (Z, individuals in columns, SNPs in rows), the physical position of each SNP (pos) and the alleles for each SNP (alleles). Load this data into the R environment.

```{r, echo=F}
load('ABO-CHB.rda')
ls()
```

2. How many individuals and how many SNPs are there in the database? What percentage of the data is missing?

```{r, echo=F}
Z = t(Z)
print(paste("Individuals ",nrow(Z)))
print(paste("SNPs ",ncol(Z)))

missing = apply(Z,2,function(x)(is.na(x)))
per_missing = sum(missing)/length(missing)
print(paste("Percentage of data missing", per_missing * 100))
```

3. Depict all SNPs simultaeneously in a ternary plot, and comment on your result. Do you believe Hardy-Weinberg equilibrium is tenable for the markers in this database?

```{r}
snpg <- genotype(Z,sep="")
geno_count = MakeCounts(Z,alleles)
ug <- UniqueGenotypeCounts(unlist(geno_count[,1:3]),verbose=F)
ternary <- HWTernaryPlot(ug[,c(1:3)])
```

We see that most of the observation are in equilibrium. There are 2 observations that are significant. 5% of the observations might lost by chance (45 * 0.05). Yes I belive Hardy-Weinberg equilibrium is tenable.

4. Using the function LD from the genetics package, compute the LD statistic D for the first two SNPs in the database. Is there significant association between these two SNPs?

```{r}
SNP1 = genotype(Z[,1],sep="")
SNP2 = genotype(Z[,2],sep="")
snp12 = data.frame(SNP1,SNP2)
LD(snp12)
```

We accept the Ho that there is no significant assoiciation. The correlation is low and D' is close to 1 which is counter intutive to the other statistics. P value and Chi Square values show that values are high and ther is no assoiciation.

5. Given your previous estimate of D, and using the formulas from the lecture slides, compute the statistics *D', χ2, R2* and *r* by hand for the first pair of SNPs. Do your results coincide with those obtained by the LD function? Can you explain possible differences?
```{r}
D = -0.0160086
PA = 80/90 ; Pa = 10/90
PB = 77/90 ; Pb = 77/90

ht = rbind(c( PA*PB , PA*Pa, PA),c(Pa*PB, Pa*Pb, Pa),c(PB,Pb ,1))
rownames(ht) = c("A","a","")
colnames(ht) = c("B","b","")

#D'
if(D>0){
  D_p = min(PA*Pb,Pa*PB)
} else{
  D_p = min(PA*PB,Pa*Pb)
}

PAB = D + PA*PB
D = PAB - PA*PB

n_indiv = 45
R_sq = D^2/(PA*PB*Pa*Pb)
X_sq = R_sq*2*n_indiv
r = sqrt(R_sq)*D/abs(D)

paste("D'=",D_p, "R-squared =",R_sq, "X-squared =", X_sq, "r =", r)
```

Some results concide like R-squared, but our case D' is different. `r` and `R^2` are pretty much the same, there is no significant *linkage desequilibrium*.

6. Given your previous estimate of D, infer the haplotype frequencies. Which haplotype is the most common?
```{r}
htd= ht + matrix(c(D,-D,0,-D,D,0,0,0,0),nrow = 3, ncol=3, byrow = T)
htd
max(htd[1:2,1:2])
```

The largest probability is for the haplotype `AB` with a probability of `r max(htd[1:2,1:2])`. 

7. Compute 4 LD statistics for all the marker pairs in this data base (D, D0, χ2 and R2). Make a scatterplot matrix of these. Is there an exact linear relationship between χ2 and R2? Why (not) so?

```{r}
geno <- function(x){
  return(genotype(x, sep=""))
}
gt = apply(Z,2,geno)
test = LD(makeGenotypes(gt))
vecD <- test$D[upper.tri(test$D)]
vecD_p <- test$`D'`[upper.tri(test$`D'`)]
vec_chi <- test$`X^2`[upper.tri(test$`X^2`)]
vec_rsq <- test$`R^2`[upper.tri(test$`R^2`)]
vectors <- as.data.frame(cbind("D" = vecD, "D'" = vecD_p, "X^2" = vec_chi, "R^2" = vec_rsq))
plot(vectors)
```

Yes, there is a linear relationship between X^2 and R2. Both ChiSquare and R square are used to evalute the goodness of fit hence they are correlated.

8. Compute a distance matrix with the distance in base pairs between all possible pairs of SNPs. Make a plot of the R2 statistics against the distance between the markers. Comment on your results.

```{r, echo=F}
plot(vectors$`R^2`, matrix(dist(pos)), main = "Correlation of Markers")
```

There seems to be no relation between R square and the SNPs. They seem to be randomly distributed. The observations are slightly dense around the bottom right.

9. Make two LD heatmaps of the markers in this database, one using the R2 statistic and one using the D' statistic, and use the positional information on the markers. Are the results consistent?

```{r, echo=F}
LDheatmap(test$`R^2`)
LDheatmap(test$`D'`)
```

Both the heatmaps produce differnt results. The heatmap with R2 has lower correlations compared to heatmap with D' statistic. The results dont seem to be consistent.

10. Simulate 45 independent SNPs under the assumption of Hardy-Weinberg equilibrium. Simulate as many SNPs as you have in your database, and take care to match each SNP in your database with a simulated SNP that has the same sample size and allele frequency. Make two LD heatmaps of the simulated SNPs, one using R2 and one using D0. Compare these to the LD heatmap of the ABO region. What do you observe? State your conclusions.

```{r, echo=F, warning=F}

make_geno =function(k){
  geno= c()
  for(i in 1:3){
    geno_name = paste(unlist(strsplit(names(k)[i], "")), collapse = '/')
    rep_geno = rep(geno_name,k[i])
    
    geno = c(geno,rep_geno)
  }
  geno_sample = sample(geno)
  return(geno_sample)
}
sample_Z = HWData(45,n=45)
gen_sample = apply(sample_Z,1,make_geno)
test1 = LD(makeGenotypes(gen_sample))
LDheatmap(test1$`R^2`)
LDheatmap(test1$`D'`)
```

We see comparatively less assoiciations in the simulated LD heatmap. Since the data is randomly geneated it fails to capture the infromations in the marker and the correlations. The values also seem to be lower.

11. Do you think there is strong or weak LD for the ABO region you just studied? Explain your opinion.

There is a strong link for the linkage disequilibrium for some metrics like D statistics. We see a lot of assoiciations between the snps.