---
title: "Assignment 6"
author: "Krishna Kalyan"
date: "1/3/2017"
output: html_document
---

```{r, echo = F, warning=FALSE,  message=FALSE}
rm(list = ls())
library(genetics)
library(knitr)
setwd('/Users/krishna/MIRI/BSG/R_BSG/Assignment6')
load('MEX_chr22.rda')
```

1. Download the file MEX_chr22.rda from the website of the course. This file contains genotype information of 76 individuals from the Mexican population. The genotype information concerns 20,531 SNPs on chromosome 22. Load this data into the R environment. The data file contains a matrix X.Geno containing the allele counts (0, 1 or 2) for one of the alleles of each SNP. The data file also provides a matrix X.Fam with pedigree information?.

```{r}
X <- t(X.Geno)
```

2. What percentage of the data is missing?

```{r}
missing <- 100 * sum(is.na(X))/length(X)
print(paste("Percentage Missing", round(missing,4)))
```

3. Plot the percentage of missing values per SNP. How many SNPs have a percentage larger than 1%?

```{r, echo=F}
nas_col <- colMeans(is.na(X))
missing_table <- data.frame(table(nas_col))
names(missing_table) <- c('percentage', 'no of snps')
kable(missing_table, digits=3)
paste("SNPS have more than 1% missing data",sum(nas_col > 0.01))
plot(nas_col)
```

4. Compute the histogram of the minor allele frequency per SNP. How many markers have minor allele frequency larger than 0.40? How many smaller than 0.05?

```{r, echo = F}
maf <- function(x) {
  y0 <-  sum(x ==0,na.rm=TRUE)
  y1 <-  sum(x ==1,na.rm=TRUE)
  y2 <-  sum(x ==2,na.rm=TRUE)
  af1 <-(100*(min(y0,y1,y2))/length(x))
  af1[af1==1] <- 0 # monomorfics
  return (af1)
}

maf_snp  <- apply(X,2,maf)
hist(maf_snp)

cond1 <- sum(maf_snp>0.40)
cond2 <- sum(maf_snp<0.05)
print(paste("snps > 0.40 ",cond1))
print(paste("snps < 0.05 ",cond2))
```

5. Consider SNPs with minor allele frequency larger than 0.40. Compute the mean (m) and the standard deviation (s) of the shared IBS alleles for all pairs of individuals. Plot against m. Use the pedigree information to identify parent-offspring (PO) pairs. Do you think that there is a clear separation between PO and the rest of pairs of individuals? Comment on your findings.

```{r, echo =F}
snps_maf4 <- X[,which(maf_snp > 0.4)]

ibs.mean <- function(x,y) {
  y <- mean(2 - abs(x - y),na.rm=TRUE)
  return(y)
}

ibs.sd <- function(x,y) {
  y <- sd(abs(x-y),na.rm=TRUE)
  return(y)
}

n <- nrow(X.Fam)
Dmean <- matrix(NA,nrow=n,ncol=n)
Dsd <- matrix(NA,nrow=n,ncol=n)
for(i in 1:n) {
  for(j in 1:n) {
    Dmean[i,j] <- ibs.mean(snps_maf4[i,],snps_maf4[j,])
    Dsd[i,j] <- ibs.sd(snps_maf4[i,],snps_maf4[j,])
    }
  }
ibs.m <- Dmean[lower.tri(Dmean)]
ibs.s <- Dsd[lower.tri(Dsd)]
is.po <- function(ind1, ind2) {
  # ind1 child of ind2
  ind1.child.ind2 <- ind1[1]==ind2[1] & (ind1[3]==ind2[2] | ind1[4]==ind2[2])
  ind1.parent.ind2 <- ind1[1]==ind2[1] & (ind1[2]==ind2[3] | ind1[2]==ind2[4]) # x     parent of y
  rel <- ind1.child.ind2 | ind1.parent.ind2
  return(rel)
}

Relationship <- matrix(NA,nrow=n,ncol=n)
  for(i in 1:n) {
    for(j in 1:n) {
      Relationship[i,j] <- is.po(X.Fam[i,],X.Fam[j,])
    }
}

rel.ind <- Relationship[lower.tri(Relationship)]
cols.rel.ind <- rep('black', length(rel.ind))
cols.rel.ind[rel.ind] <- 'red'
plot(ibs.m, ibs.s, xlim=c(1,2), ylim=c(0,1),
main = 'SNPs with MAF >= 0.4',col = cols.rel.ind,
xlab="Mean", ylab="Standard deviation")

print("There is a separation between PO and the rest of pairs of individuals. We can clearly see the formation of clusters.")
```

6. Repeat the question 5 considering SNPs with minor allele frequency smaller than 0.05. Do you think that there is a clear separation between PO and the rest of pairs of individuals? Comment on your findings.

```{r, echo =F}
snps_maf5 <- X[,which(maf_snp < 0.05)]

ibs.mean <- function(x,y) {
  y <- mean(2 - abs(x - y),na.rm=TRUE)
  return(y)
}

ibs.sd <- function(x,y) {
  y <- sd(abs(x-y),na.rm=TRUE)
  return(y)
}

n <- nrow(X.Fam)
Dmean <- matrix(NA,nrow=n,ncol=n)
Dsd <- matrix(NA,nrow=n,ncol=n)
for(i in 1:n) {
  for(j in 1:n) {
    Dmean[i,j] <- ibs.mean(snps_maf5[i,],snps_maf5[j,])
    Dsd[i,j] <- ibs.sd(snps_maf5[i,],snps_maf5[j,])
    }
  }
ibs.m <- Dmean[lower.tri(Dmean)]
ibs.s <- Dsd[lower.tri(Dsd)]
is.po <- function(ind1, ind2) {
  # ind1 child of ind2
  ind1.child.ind2 <- ind1[1]==ind2[1] & (ind1[3]==ind2[2] | ind1[4]==ind2[2])
  ind1.parent.ind2 <- ind1[1]==ind2[1] & (ind1[2]==ind2[3] | ind1[2]==ind2[4]) # x     parent of y
  rel <- ind1.child.ind2 | ind1.parent.ind2
  return(rel)
}

Relationship <- matrix(NA,nrow=n,ncol=n)
  for(i in 1:n) {
    for(j in 1:n) {
      Relationship[i,j] <- is.po(X.Fam[i,],X.Fam[j,])
    }
}

rel.ind <- Relationship[lower.tri(Relationship)]
cols.rel.ind <- rep('black', length(rel.ind))
cols.rel.ind[rel.ind] <- 'red'
plot(ibs.m, ibs.s, xlim=c(1,2), ylim=c(0,1),
main = 'SNPs with MAF < 0.5',col = cols.rel.ind,
xlab="Mean", ylab="Standard deviation")

print("This time the seperation is not too clear. The markers have very low MAF and seem to mostly comprise of monomers ")
```

7. Consider again SNPs with minor allele frequency larger than 0.40. Compute and plot
the fraction of loci sharing 0 against the fraction of loci sharing 2 IBS alleles for all pairs of individuals. Use the pedigree information to identify PO pairs. Do you think that there is a clear separation between PO and the rest of pairs of individuals? Comment on your findings.

```{r, echo=F}
m = maf_snp[maf_snp > 0.4]
d = names(m)
Y.Geno =X.Geno[ (rownames(X.Geno) %in% d), ]
s.0 <- function(x,y) {
  return(sum(x==0 & y==0,na.rm=TRUE))
}

s.1 <- function(x,y) {
  return(sum(x==1 & y==1,na.rm=TRUE))
}
s.2 <- function(x,y) {
  return(sum(x==2 & y==2,na.rm=TRUE))
}

s0 <- matrix(NA,nrow=n,ncol=n)
s1 <- matrix(NA,nrow=n,ncol=n)
s2 <- matrix(NA,nrow=n,ncol=n)

for(i in 1:n) {
  for(j in 1:n) {
    s0[i,j] <- s.0(Y.Geno[,i],Y.Geno[,j])
    s1[i,j] <- s.1(Y.Geno[,i],Y.Geno[,j])
    s2[i,j] <- s.2(Y.Geno[,i],Y.Geno[,j])
  }
}
ibs.m <- s0[lower.tri(s0)]
ibs.s <- s2[lower.tri(s2)]
plot(ibs.m,ibs.s,xlab="Fraccion of loci with 0 IBS alleles shared (p0)",ylab="Fraccion of loci with 2 IBS alleles shared (p2)", col=cols.rel.ind)
print("P0 pairs have more loci sharing 2")
```
8. Repeat the question 7 considering SNPs with minor allele frequency smaller than 0.05. Do you think that there is a clear separation between PO and the rest of pairs of individuals? Comment on your findings.

```{r, echo=F}
m = maf_snp[maf_snp < 0.05]
d = names(m)
Y.Geno =X.Geno[ (rownames(X.Geno) %in% d), ]
s.0 <- function(x,y) {
  return(sum(x==0 & y==0,na.rm=TRUE))
}

s.1 <- function(x,y) {
  return(sum(x==1 & y==1,na.rm=TRUE))
}
s.2 <- function(x,y) {
  return(sum(x==2 & y==2,na.rm=TRUE))
}

s0 <- matrix(NA,nrow=n,ncol=n)
s1 <- matrix(NA,nrow=n,ncol=n)
s2 <- matrix(NA,nrow=n,ncol=n)

for(i in 1:n) {
  for(j in 1:n) {
    s0[i,j] <- s.0(Y.Geno[,i],Y.Geno[,j])
    s1[i,j] <- s.1(Y.Geno[,i],Y.Geno[,j])
    s2[i,j] <- s.2(Y.Geno[,i],Y.Geno[,j])
  }
}
ibs.m <- s0[lower.tri(s0)]
ibs.s <- s2[lower.tri(s2)]
plot(ibs.m,ibs.s,xlab="Fraccion of loci with 0 IBS alleles shared (p0)",ylab="Fraccion of loci with 2 IBS alleles shared (p2)", col=cols.rel.ind)
print("The is a partial seperation between PO")
```